sequelize/umzug
# feat: non-breaking ESM support 🤝

Closes #608

This adds some esm-vs-commonjs "smartness" to umzug:

1. use `require` for `.cjs` migrations and `import` for `.mjs` migrations (and their typescript equivalents)
2. use `require` for `.js` migrations _if_ `typeof require.main === 'object'`, and `import` to `.js` migrations otherwise
3. use the same criteria to create (c)js vs mjs templates when creating migration files
4. add `"moduleResolution": "Node16"` to tsconfig.lib.json to make sure `import(filepath)` doesn't get transpiled into `__importStar(require(filepath))` (see [here](https://github.com/microsoft/TypeScript/issues/43329#issuecomment-926931937) and [here](https://github.com/microsoft/TypeScript/issues/43329#issuecomment-1276000768))

Tests/examples:

- add a `vanilla-esm` example to make sure using `import` / top-level await works
- add a step to the `test_pkg` job to make sure vitest isn't hiding gnarly import problems - this is installing the compiled library as a `.tgz`, and with no other dev/prod dependencies like vitest or ts-node having been installed, so should be very close to what end users will do

Didn't:

- add a wrapper.mjs file in the compiled folder as suggested in https://github.com/sequelize/umzug/issues/608#issuecomment-1611845497, mostly just because it didn't seem to be necessary? It seems to work fine when imported from an ES-module, using top-level await, etc., even though umzug is itself a commonjs module.

<details>
<summary>original body</summary>

~Related to #608 - although does not close it.~

~This adds built-in support for `.mjs` and `.mts` files. `.mjs` should "just work" - write migrations as ESM modules and they'll be imported in the right way. For the current major version, `.js` will continue to be assumed commonjs. ESM-fans will just have to type that extra `m` in their filenames.~

~This PR _doesn't_ add a wrapper file so that the umzug library itself can be imported as an ES module. That can be done in a follow-up PR. In the meantime, ESM users can use `createRequire` as in the [existing ESM example](https://github.com/sequelize/umzug/tree/main/examples/2.es-modules).~
</details>
